File: php/v2/examples/ListBrowserTargetingOptions.php
<?php
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Require the base class.
require_once __DIR__ . "/../BaseExample.php";

/**
 * This example lists all of the targeting options available for the Browser
 * targeting type.
 */
class ListBrowserTargetingOptions extends BaseExample
{
    /**
     * (non-PHPdoc)
     * @see BaseExample::getInputParameters()
     */
    protected function getInputParameters(): array
    {
        return array(
            array(
                'name' => 'advertiser_id',
                'display' => 'Advertiser ID',
                'required' => true
            )
        );
    }

    /**
     * (non-PHPdoc)
     * @see BaseExample::run()
     */
    public function run()
    {
        $values = $this->formValues;
        $advertiserId = $values['advertiser_id'];

        print '<h2>Listing all available browser targeting options</h2>';

        $response = null;
        $nextPageToken = null;

        do {
            $optParams = array(
                'advertiserId' => $advertiserId,
                'pageToken' => $nextPageToken
            );

            // Call the API, getting the browser targeting options for the
            // identified advertiser.
            $response = $this
                ->service
                ->targetingTypes_targetingOptions
                ->listTargetingTypesTargetingOptions(
                    'TARGETING_TYPE_BROWSER',
                    $optParams
                );

            if (!empty($response->getTargetingOptions())) {
                $this->printTargetingOptions($response->getTargetingOptions());
            } else {
                print '<p>No targeting options returned</p>';
            }

            // Update the next page token.
            $nextPageToken = $response->getNextPageToken();
        } while (
            !empty($response->getTargetingOptions()) &&
            !empty($nextPageToken)
        );
    }

    /**
     * (non-PHPdoc)
     * @see BaseExample::getName()
     */
    public static function getName(): string
    {
        return 'List Browser Targeting Options';
    }

    /**
     * Prints the given browser targeting options.
     * @param array $options browser targeting options to be printed.
     */
    protected function printTargetingOptions(array $options)
    {
        foreach ($options as $option) {
            printf(
                '<p>Targeting Option ID: %s, Browser Display Name: %s</p>',
                $option['targetingOptionId'],
                $option['browserDetails']['displayName']
            );
        }
    }
}

File: python/v3/list_browser_targeting_options.py
#!/usr/bin/python
#
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""This example lists browser targeting options for the given advertiser."""

import argparse
import os
import sys

from googleapiclient.errors import HttpError

sys.path.insert(0, os.path.abspath(".."))
import samples_util


def main(service, flags):
    """Retrieve available browser targeting options.

    Args:
      service: the googleapiclient.discovery.Resource instance used to interact
        with the Display & Video 360 API.
      flags: the parsed command-line arguments.

    Raises:
      HttpError: If an API request is not made successfully.
    """

    # Create the page token variable.
    next_page_token = ""

    while True:
        # Request the targeting options list.
        response = (
            service.targetingTypes()
            .targetingOptions()
            .list(
                advertiserId=flags.advertiser_id,
                targetingType="TARGETING_TYPE_BROWSER",
                pageToken=next_page_token,
            )
            .execute()
        )

        # Check if response is empty.
        if response:
            # Iterate over retrieved targeting options.
            for option in response["targetingOptions"]:
                print(
                    f"Targeting Option ID: {option['targetingOptionId']}, "
                    "Browser Display Name: "
                    f"{option['browserDetails']['displayName']}"
                )
        else:
            print("List request returned no Targeting Options.")
            break

        # Update the next page token.
        # Break out of loop if there is no next page.
        if "nextPageToken" in response:
            next_page_token = response["nextPageToken"]
        else:
            break


if __name__ == "__main__":
    # Declare command-line flags.
    argparser = argparse.ArgumentParser(add_help=False)
    argparser.add_argument(
        "advertiser_id",
        help="The ID of the advertiser this request is being made within.",
    )

    # Retrieve command line arguments.
    flags = samples_util.get_arguments(
        sys.argv,
        __doc__,
        parents=[samples_util.get_default_parser(), argparser],
    )

    # Authenticate and construct service.
    service = samples_util.get_service(
        version="v3", useServiceAccount=flags.use_service_account
    )

    try:
        main(service, flags)
    except HttpError as e:
        print(e)
        sys.exit(1)
